src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -198,820 +542,1177 @@ private void checkDataLength(int ... lengths) {
+                    src.remaining(), dst.array(),
+                    dst.arrayOffset() + dst.position());
+                src.position(src.limit());
+                dst.position(dst.position() + len);
+                return len;
+            }
+
+            // Check for overlap in the bytebuffers
+            dst = overlapDetection(src, dst);
 
-            // Get the current memory address for the given ByteBuffers
-            long srcaddr = dsrc.address();
-            long dstaddr = ddst.address();
+            // Length of the input
+            ByteBuffer tag;
+            ByteBuffer ct = src.duplicate();
 
-            // Find the lowest attachment that is the base memory address of the
-            // shared memory for the src object
-            while (dsrc.attachment() != null) {
-                srcaddr = ((DirectBuffer) dsrc.attachment()).address();
-                dsrc = (DirectBuffer) dsrc.attachment();
+            ByteBuffer buffer = ((ibuffer == null || ibuffer.size() == 0) ?
+                null : ByteBuffer.wrap(ibuffer.toByteArray()));
+            int len;
+
+            if (ct.remaining() >= tagLenBytes) {
+                tag = src.duplicate();
+                tag.position(ct.limit() - tagLenBytes);
+                ct.limit(ct.limit() - tagLenBytes);
+                len = ct.remaining();
+                if (buffer != null) {
+                    len += buffer.remaining();
+                }
+            } else if (buffer != null) {
+                // It's unlikely the tag will be between the buffer and data
+                tag = ByteBuffer.allocate(tagLenBytes);
+                int limit = buffer.remaining() - (tagLenBytes - ct.remaining());
+                buffer.mark();
+                buffer.position(limit);
+                // Read from "new" limit to buffer's end
+                tag.put(buffer);
+                // reset buffer to data only
+                buffer.reset();
+                buffer.limit(limit);
+                tag.put(ct);
+                tag.flip();
+                // Limit is how much of the ibuffer has been chopped off.
+                len = buffer.remaining();
+            } else {
+                throw new AEADBadTagException("Input too short - need tag");
             }
 
-            // Find the lowest attachment that is the base memory address of the
-            // shared memory for the dst object
-            while (ddst.attachment() != null) {
-                dstaddr = ((DirectBuffer) ddst.attachment()).address();
-                ddst = (DirectBuffer) ddst.attachment();
+            // 'len' contains the length in ibuffer and src
+            checkDataLength(len);
+
+            // Save GHASH context to allow the tag to be checked even though
+            // the dst buffer is too short.  Context will be restored so the
+            // method can be called again with the proper sized dst buffer.
+            if (len > dst.remaining()) {
```

Comment:
We should check the dst capacity in the beginning of the method, if its capacity is too small, i.e. less than the overall length - tag length, don't proceed further.
As in the doFinal() w/ byte[] arguments, I am not sure if the save/restore is really necessary.
Maybe add a test if it's not already covered by existing tests if time permits.

Reply:
I think it can be moved up without doing any extra work
I'm pretty sure there is a test because I remember Sean C working on a bug with restoring when dst was too small.